### PR TITLE
Mccalluc/better ids

### PR DESCRIPTION
**Pull request by mccalluc**
This provides more informative row labels in the heatmap. I have a few concerns:
- Search is just doing a linear scan of the labels: if the labels get longer, search will slow down proportionally. I had tried using Whoosh, but the indexing took a very long time. Might be worth trying to find an intermediate solution.
- Manually testing it, with some frequency, things just don't work. I've thought these were race conditions in Dash that were outside my control, but I'm not really confident about that. With these changes, I feel like it's even more frequent. Maybe we should understand those before pushing the framework even harder.
- Estimating the width of the labels: Right now I'm just assuming 10 pixels per letter, but if we're putting more in there, that estimate could be wide of the mark. Is it worth getting actual character widths?

I had begun it thinking it'd be an easy way to get a new version out, but there are enough changes here that I don't think that would be a good idea. I'll go back and make a trivial branch, and hopefully the travis rebuild will work.

### CODE REVIEW

**Comment by mccalluc** [comment]
- table not working on demo data
- ok to do left substring search, but could be a later pr
- there may be a lot of text columns: just take the last one

File: context/tests/test_frames.py [hunk 1]
```diff
@@ -5,15 +5,16 @@
 import numpy as np
```
Comment (scottx611x) [inline context/tests/test_frames.py]: Is the functionality being tested here what you'd be pulling out into a common PyPI package?

Comment (mccalluc) [reply context/tests/test_frames.py]: yeah: ` TestTabularParser` would move out. But that refactoring doesn't block this.

**Comment by gmnelson** [comment]
@mccalluc If the 2 visualizations were named sufficiently different so that users were not confused as to which to use, I don't foresee this presenting an issue right now, given the limited usage. I can discuss this more, if needed, after the meeting tomorrow.

File: context/app/utils/tabular.py [hunk 2]
```diff
@@ -4,7 +4,15 @@
...
 
 
-def parse(file, col_zero_index=True, keep_strings=False):
+def parse(file, col_zero_index=True, keep_strings=False, label_delim=None):
+    '''
+    Given a file handle, try to determine its format and return a dataframe.
+    :param file:
+    :param col_zero_index:
+    :param keep_strings: Preserve string values in dataframe? Default False.
+    :param label_delim: Concatenate string values in table? Default None.
```
Comment (jkmarx) [inline context/app/utils/tabular.py]: why the ?

File: context/app/utils/tabular.py [hunk 3]
```diff
@@ -47,4 +55,20 @@ def parse(file, col_zero_index=True, keep_strings=False):
...
+    if label_delim:
+        label_map = {
+            i: label_delim.join(
+                dataframe.select_dtypes(['object']).loc[i].tolist()
+                + [str(i)]
+            )
+            for i in dataframe.index
+        }
+    else:
+        # This is unused, but only returning
```
Comment (jkmarx) [inline context/app/utils/tabular.py]: If it's unused, do we need the else?

File: context/app/utils/tabular.py [hunk 4]
```diff
@@ -47,4 +55,20 @@ def parse(file, col_zero_index=True, keep_strings=False):
         if is_gct:
             dataframe.drop(columns=['Description'], inplace=True)
             # TODO: Combine the first two columns?
-    return dataframe if keep_strings else dataframe.select_dtypes(['number'])
+
+    if label_delim:
```
Comment (jkmarx) [inline context/app/utils/tabular.py]: Add comment about what's going on here. (a lot of chaining)

File: context/app/utils/tabular.py (hunk 2 above)
Comment (mccalluc) [reply context/app/utils/tabular.py]: The intention was that the Boolean value answers this question, but sounds like it's confusing, so I'll change it.

File: context/app/utils/tabular.py (hunk 3 above)
Comment (mccalluc) [reply context/app/utils/tabular.py]: Good point.

File: context/app/utils/tabular.py (hunk 4 above)
Comment (mccalluc) [reply context/app/utils/tabular.py]: Sure. Thanks!

**Pull request merged**